lib/policy_machine_storage_adapters/active_record/postgresql.rb

```diff
@@ -13,34 +13,84 @@ def self.transitive_closure?(ancestor, descendant)
       end
 
       def self.descendants_of(element_or_scope)
-        #FIXME: Preloading with to_a seems to be necessary because putting complex sql in start_with can
-        # lead to degenerate performance (noticed in ancestors_of call in accessible_objects)
-        # Ideally, fix the SQL so it's both a single call and performant
-        element_or_scope = [*element_or_scope]
-        case element_or_scope.size
-        when 0
-          PolicyElement.none
-        when 1
-          PolicyElement.where('"policy_elements"."id" IN (SELECT assignments__recursive.child_id FROM (WITH RECURSIVE "assignments__recursive" AS ( SELECT "assignments"."id", "assignments"."child_id", "assignments"."parent_id" FROM "assignments" WHERE "assignments"."parent_id" = ? UNION ALL SELECT "assignments"."id", "assignments"."child_id", "assignments"."parent_id" FROM "assignments" INNER JOIN "assignments__recursive" ON "assignments__recursive"."child_id" = "assignments"."parent_id" ) SELECT "assignments__recursive".* FROM "assignments__recursive") AS "assignments__recursive")', element_or_scope.first.id)
-        else
-          PolicyElement.where('"policy_elements"."id" IN (SELECT assignments__recursive.child_id FROM (WITH RECURSIVE "assignments__recursive" AS ( SELECT "assignments"."id", "assignments"."child_id", "assignments"."parent_id" FROM "assignments" WHERE "assignments"."parent_id" in (?) UNION ALL SELECT "assignments"."id", "assignments"."child_id", "assignments"."parent_id" FROM "assignments" INNER JOIN "assignments__recursive" ON "assignments__recursive"."child_id" = "assignments"."parent_id" ) SELECT "assignments__recursive".* FROM "assignments__recursive") AS "assignments__recursive")', element_or_scope.map(&:id))
-        end
+        PolicyElement.where('
+          id IN (
+            WITH RECURSIVE assignments_recursive AS (
+              (
+                SELECT child_id, parent_id
+                FROM assignments
+                WHERE parent_id in (?)
+              )
+              UNION ALL
+              (
+                SELECT assignments.child_id, assignments.parent_id
+                FROM assignments
+                INNER JOIN assignments_recursive
+                ON assignments_recursive.child_id = assignments.parent_id
+              )
+            )
+
+            SELECT assignments_recursive.child_id
+            FROM assignments_recursive
+          )',
+          [*element_or_scope].map(&:id))
```

Comment:
For such a large query, may be _slightly_ more clear to knock the final parenthesis to the next line?